### PR TITLE
feature(davide): GTrXL implementation

### PR DESCRIPTION
**Pull request by davide97l**
## Description
Implementation of the model GTrXL based on Transformer for RL.
Reference: https://arxiv.org/abs/1910.06764

### CODE REVIEW

File: ding/torch_utils/network/gtrxl.py [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+        """
+        Overview:
+            Init memory with an input list of tensors or create it automatically given its dimensions.
+        Arguments:
+            - memory: (:obj:`Optional[torch.Tensor]`): memory input.
+            Shape is (layer_num, memory_len, bs, embedding_dim).
+            memory_len is length of memory, bs is batch size and embedding_dim is the dimension of embedding.
+        """
+        if memory is not None:
+            self.memory = memory
```
Comment (PaParaZz1) [inline ding/torch_utils/network/gtrxl.py:119]: `layer_num_plus1, self.memory_len, self.bs, self.embedding_dim = memory.shape`
`self.layer_num = layer_num_plus1 - 1`

File: ding/torch_utils/network/gtrxl.py [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+                m20 m21 m22                               h10 h11 h12
+        Arguments:
+            - hidden_state: (:obj:`List[torch.Tensor]`): hidden states to update the memory.
+            Shape is (cur_seq, bs, embedding_dim) for each layer. cur_seq is length of sequence.
+        Returns:
+            - memory: (:obj:`Optional[torch.Tensor]`): output memory.
+            Shape is (layer_num, memory_len, bs, embedding_dim).
+        """
+        if self.memory is None or hidden_state is None:
+            return None
```
Comment (PaParaZz1) [inline ding/torch_utils/network/gtrxl.py:147]: why not raise exception

File: ding/torch_utils/network/gtrxl.py [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+            (
+                (query.view(cur_seq, bs, self.head_num, self.head_dim) + v),
+                r.view(cur_seq + prev_seq, self.head_num, self.head_dim),
+            ),
+        )  # cur_seq x full_seq x bs x head_num
+        position_attn = self._rel_shift(position_attn)
+        attn = content_attn + position_attn  # cur_seq x full_seq x bs x head_num
+        attn.mul_(self.scale)
+
+        if mask is not None and mask.any().item():
```
Comment (PaParaZz1) [inline ding/torch_utils/network/gtrxl.py:276]: check the shape of mask

File: ding/torch_utils/network/gtrxl.py [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+            - head_num (:obj:`int`): number of heads for multihead attention
+            - mlp_num (:obj:`int`): number of mlp layers in attention layer
+            - dropout (:obj:`nn.Module`): dropout
+            - activation (:obj:`nn.Module`): activation function
+            - gating (:obj:`bool`): whether to use gating mechanism or not
+        """
+        super(GatedTransformerXLLayer, self).__init__()
+        self.dropout = dropout
+        self.gating = gating
+        self.gate1 = GRUGatingUnit(input_dim)
```
Comment (PaParaZz1) [inline ding/torch_utils/network/gtrxl.py:326]: only enable `GRUGatingUnit` when `self.gating is True`

File: ding/torch_utils/network/gtrxl.py [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+        layers = []
+        dims = [input_dim] + [hidden_dim] * (mlp_num - 1) + [input_dim]
+        for i in range(mlp_num):
+            layers.append(fc_block(dims[i], dims[i + 1], activation=activation))
+            if i != mlp_num - 1:
+                layers.append(self.dropout)
+        layers.append(self.dropout)
+        self.mlp = nn.Sequential(*layers)
+        self.layernorm1 = build_normalization('LN')(input_dim)
+        self.layernorm1 = build_normalization('LN')(input_dim)
```
Comment (PaParaZz1) [inline ding/torch_utils/network/gtrxl.py:343]: two same layernorm, you need two different layernorm

File: ding/torch_utils/network/gtrxl.py [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+        x = self.dropout(self.embedding(x))
+        prev_seq = self.memory_len
+        full_seq = cur_seq + prev_seq
+
+        # TODO: add padding to attention mask, https://huggingface.co/docs/transformers/preprocessing
+        dec_attn_mask = (
+            torch.triu(
+                torch.ones((cur_seq, cur_seq + prev_seq)),
+                diagonal=1 + prev_seq,
+            ).bool()[..., None].to(x.device)
```
Comment (PaParaZz1) [inline ding/torch_utils/network/gtrxl.py:492]: use `unsqueeze`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/torch_utils/network/gtrxl.py [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+        r"""
+        Overview:
+            Clear or set the memory of GTrXL.
+         Arguments:
+            - batch_size (:obj:`int`): batch size
+            - state (:obj:`Optional[torch.Tensor]`): input memory.
+            Shape is (layer_num, memory_len, bs, embedding_dim).
+        """
+        self.memory = Memory()
+        if batch_size is not None:
```
[the remark this answers is not in the document]
Comment (davide97l) [reply ding/torch_utils/network/gtrxl.py:450]: We need to check batch size because it could happen that an input x has a different batch size than the batch size of the existing memory m. This could happen when we switch from collect to train phase for example. In this case the memory will be reset. In a next version, each phase could have its own memory with a wrapper so to avoid this problem.

File: ding/torch_utils/network/gtrxl.py [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+        inv_freq = 1 / (10000 ** (torch.arange(0.0, embedding_dim, 2.0) / embedding_dim))  # (embedding_dim / 2.0)
+        self.register_buffer('inv_freq', inv_freq)
+
+    def forward(self, pos_seq):
+        """
+        Overview:
+            Compute positional embedding
+        Arguments:
+            - pos_seq: (:obj:`torch.Tensor`): positional sequence,
+             usually a 1D integer sequence as [seq_len-1, seq_len-2, ..., 1, 0],
```
[the remark this answers is not in the document]
Comment (davide97l) [reply ding/torch_utils/network/gtrxl.py:30]: I think here the order is the same, I found both orders in different Transformer implementations

File: ding/policy/r2d2_gtrxl.py [hunk 9]
```diff
@@ -0,0 +1,583 @@
...
+        self._priority = self._cfg.priority
+        self._priority_IS_weight = self._cfg.priority_IS_weight
+        self._optimizer = Adam(self._model.parameters(), lr=self._cfg.learn.learning_rate)
+        self._gamma = self._cfg.discount_factor
+        self._nstep = self._cfg.nstep
+        self._burnin_step = self._cfg.burnin_step
+        self._batch_size = self._cfg.learn.batch_size
+        self._seq_len = self._cfg.seq_len
+        self._value_rescale = self._cfg.learn.value_rescale
+        self._init_memory = self._cfg.learn.init_memory
```
Comment (PaParaZz1) [inline ding/policy/r2d2_gtrxl.py:263]: check the value of `self._init_memory` in proper range

File: ding/policy/r2d2_gtrxl.py [hunk 10]
```diff
@@ -0,0 +1,583 @@
...
+    def _reset_learn(self, data_id: Optional[List[int]] = None) -> None:
+        self._learn_model.reset(data_id=data_id)
+        self._target_model.reset(data_id=data_id)
+        self._learn_model.reset_memory()
+        self._target_model.reset_memory()
+
+    def _state_dict_learn(self) -> Dict[str, Any]:
+        return {
+            'model': self._learn_model.state_dict(),
+            'optimizer': self._optimizer.state_dict(),
```
Comment (PaParaZz1) [inline ding/policy/r2d2_gtrxl.py:448]: save `target_model` in state_dict

File: ding/policy/r2d2_gtrxl.py [hunk 11]
```diff
@@ -0,0 +1,583 @@
...
+        r"""
+        Overview:
+            Collect mode init method. Called by ``self.__init__``.
+            Init traj and unroll length, collect model.
+        """
+        assert 'unroll_len' not in self._cfg.collect, "Use default unroll_len"
+        self._nstep = self._cfg.nstep
+        self._gamma = self._cfg.discount_factor
+        self._unroll_len = self._cfg.unroll_len
+        self._collect_model = model_wrap(self._model, seq_len=self._seq_len, wrapper_name='transformer_segment')
```
Comment (PaParaZz1) [inline ding/policy/r2d2_gtrxl.py]: whether there is some order requirements for different transformer wrapper

File: ding/policy/r2d2_gtrxl.py [hunk 12]
```diff
@@ -0,0 +1,583 @@
...
+            - timestep (:obj:`namedtuple`): Output after env step, including at least ['reward', 'done'] \
+                (here 'obs' indicates obs after env step).
+        Returns:
+            - transition (:obj:`dict`): Dict type transition data.
+        """
+        transition = {
+            'obs': obs,
+            'action': model_output['action'],
+            'prev_memory': model_output['memory'],
+            'prev_state': None,
```
Comment (PaParaZz1) [inline ding/policy/r2d2_gtrxl.py:517]: why use `prev_state` placeholder

File: ding/policy/r2d2_gtrxl.py [hunk 13]
```diff
@@ -0,0 +1,583 @@
...
+        self._eval_model.eval()
+        with torch.no_grad():
+            output = self._eval_model.forward(data)
+        if self._cuda:
+            output = to_device(output, 'cpu')
+        output = default_decollate(output)
+        return {i: d for i, d in zip(data_id, output)}
+
+    def _reset_eval(self, data_id: Optional[List[int]] = None) -> None:
+        self._eval_model.reset(data_id=data_id)
```
Comment (PaParaZz1) [inline ding/policy/r2d2_gtrxl.py:572]: why call `reset_memory` only in `forward_eval`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — APPROVED** [review]
